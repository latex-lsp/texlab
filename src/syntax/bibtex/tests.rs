use insta::assert_debug_snapshot;

use super::{parse, SyntaxNode};

fn setup(input: &str) -> SyntaxNode {
    SyntaxNode::new_root(parse(input))
}

#[test]
fn test_preamble() {
    assert_debug_snapshot!(setup(r#"@preamble{"Hello World!"}"#));
}

#[test]
fn test_comment() {
    assert_debug_snapshot!(setup(
        r#"@comment{aksin,
    author       = {Aks{\i}n, {\"O}zge and T{\"u}rkmen, Hayati and Artok, Levent
                    and {\c{C}}etinkaya, Bekir and Ni, Chaoying and
                    B{\"u}y{\"u}kg{\"u}ng{\"o}r, Orhan and {\"O}zkal, Erhan},
    title        = {Effect of immobilization on catalytic characteristics of
                    saturated {Pd-N}-heterocyclic carbenes in {Mizoroki-Heck}
                    reactions},
    journaltitle = jomch,
    date         = 2006,
    volume       = 691,
    number       = 13,
    pages        = {3027-3036},
    indextitle   = {Effect of immobilization on catalytic characteristics},
}"#
    ));
}

#[test]
fn test_entry_colon() {
    assert_debug_snapshot!(setup(
        r#"@article{foo:2019,
    author = {Foo Bar},
    title = {Baz Qux},
    year = {2019},
}"#
    ));
}

#[test]
fn test_biblatex_examples_001() {
    assert_debug_snapshot!(setup(
        r#"@incollection{westfahl:space,
    author       = {Westfahl, Gary},
    title        = {The True Frontier},
    subtitle     = {Confronting and Avoiding the Realities of Space in {American}
                    Science Fiction Films},
    pages        = {55-65},
    crossref     = {westfahl:frontier},
    langid       = {english},
    langidopts   = {variant=american},
    indextitle   = {True Frontier, The},
    annotation   = {A cross-referenced article from a \texttt{collection}. This is
                    an \texttt{incollection} entry with a \texttt{crossref}
                    field. Note the \texttt{subtitle} and \texttt{indextitle}
                    fields},
}

% booktitle and booksubtitle are only needed for BibTeX's less sophisticated
% inheritance set-up to make sure westfahl:space shows correctly.
% With Biber they are not needed.
@collection{westfahl:frontier,
    editor       = {Westfahl, Gary},
    title        = {Space and Beyond},
    date         = 2000,
    subtitle     = {The Frontier Theme in Science Fiction},
    publisher    = {Greenwood},
    location     = {Westport, Conn. and London},
    langid       = {english},
    langidopts   = {variant=american},
    booktitle    = {Space and Beyond},
    booksubtitle = {The Frontier Theme in Science Fiction},
    annotation   = {This is a \texttt{collection} entry. Note the format of the
                    \texttt{location} field as well as the \texttt{subtitle}
                    field},
}"#
    ));
}

#[test]
fn test_biblatex_examples_002() {
    assert_debug_snapshot!(setup(
        r#"@string{jomch   = {J.~Organomet. Chem.}}

@article{aksin,
    author       = {Aks{\i}n, {\"O}zge and T{\"u}rkmen, Hayati and Artok, Levent
                    and {\c{C}}etinkaya, Bekir and Ni, Chaoying and
                    B{\"u}y{\"u}kg{\"u}ng{\"o}r, Orhan and {\"O}zkal, Erhan},
    title        = {Effect of immobilization on catalytic characteristics of
                    saturated {Pd-N}-heterocyclic carbenes in {Mizoroki-Heck}
                    reactions},
    journaltitle = jomch,
    date         = 2006,
    volume       = 691,
    number       = 13,
    pages        = {3027-3036},
    indextitle   = {Effect of immobilization on catalytic characteristics},
}"#
    ));
}

#[test]
fn test_biblatex_examples_003() {
    assert_debug_snapshot!(setup(
        r#"@article{angenendt,
    author       = {Angenendt, Arnold},
    title        = {In Honore Salvatoris~-- Vom Sinn und Unsinn der
                    Patrozinienkunde},
    journaltitle = {Revue d'Histoire Eccl{\'e}siastique},
    date         = 2002,
    volume       = 97,
    pages        = {431--456, 791--823},
    langid       = {german},
    indextitle   = {In Honore Salvatoris},
    shorttitle   = {In Honore Salvatoris},
    annotation   = {A German article in a French journal. Apart from that, a
                    typical \texttt{article} entry. Note the \texttt{indextitle}
                    field},
}"#
    ));
}
