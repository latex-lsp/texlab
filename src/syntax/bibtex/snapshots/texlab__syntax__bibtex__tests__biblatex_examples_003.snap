---
source: src/syntax/bibtex/tests.rs
assertion_line: 109
expression: "setup(r#\"@article{angenendt,\n    author       = {Angenendt, Arnold},\n    title        = {In Honore Salvatoris~-- Vom Sinn und Unsinn der\n                    Patrozinienkunde},\n    journaltitle = {Revue d'Histoire Eccl{\\'e}siastique},\n    date         = 2002,\n    volume       = 97,\n    pages        = {431--456, 791--823},\n    langid       = {german},\n    indextitle   = {In Honore Salvatoris},\n    shorttitle   = {In Honore Salvatoris},\n    annotation   = {A German article in a French journal. Apart from that, a\n                    typical \\texttt{article} entry. Note the \\texttt{indextitle}\n                    field},\n}\"#)"
---
ROOT@0..616
  ENTRY@0..616
    TYPE@0..8 "@article"
    L_DELIM@8..9 "{"
    NAME@9..18 "angenendt"
    COMMA@18..19 ","
    WHITESPACE@19..24 "\n    "
    FIELD@24..59
      NAME@24..30 "author"
      WHITESPACE@30..37 "       "
      EQ@37..38 "="
      WHITESPACE@38..39 " "
      CURLY_GROUP@39..58
        L_CURLY@39..40 "{"
        WORD@40..49 "Angenendt"
        COMMA@49..50 ","
        WHITESPACE@50..51 " "
        WORD@51..57 "Arnold"
        R_CURLY@57..58 "}"
      COMMA@58..59 ","
    WHITESPACE@59..64 "\n    "
    FIELD@64..166
      NAME@64..69 "title"
      WHITESPACE@69..77 "        "
      EQ@77..78 "="
      WHITESPACE@78..79 " "
      CURLY_GROUP@79..165
        L_CURLY@79..80 "{"
        WORD@80..82 "In"
        WHITESPACE@82..83 " "
        WORD@83..89 "Honore"
        WHITESPACE@89..90 " "
        WORD@90..100 "Salvatoris"
        NBSP@100..101 "~"
        WORD@101..103 "--"
        WHITESPACE@103..104 " "
        WORD@104..107 "Vom"
        WHITESPACE@107..108 " "
        WORD@108..112 "Sinn"
        WHITESPACE@112..113 " "
        WORD@113..116 "und"
        WHITESPACE@116..117 " "
        WORD@117..123 "Unsinn"
        WHITESPACE@123..124 " "
        WORD@124..127 "der"
        WHITESPACE@127..148 "\n                    "
        WORD@148..164 "Patrozinienkunde"
        R_CURLY@164..165 "}"
      COMMA@165..166 ","
    WHITESPACE@166..171 "\n    "
    FIELD@171..224
      NAME@171..183 "journaltitle"
      WHITESPACE@183..184 " "
      EQ@184..185 "="
      WHITESPACE@185..186 " "
      CURLY_GROUP@186..223
        L_CURLY@186..187 "{"
        WORD@187..192 "Revue"
        WHITESPACE@192..193 " "
        WORD@193..203 "d'Histoire"
        WHITESPACE@203..204 " "
        WORD@204..208 "Eccl"
        CURLY_GROUP@208..213
          L_CURLY@208..209 "{"
          ACCENT@209..212
            ACCENT_NAME@209..211 "\\'"
            WORD@211..212 "e"
          R_CURLY@212..213 "}"
        WORD@213..222 "siastique"
        R_CURLY@222..223 "}"
      COMMA@223..224 ","
    WHITESPACE@224..229 "\n    "
    FIELD@229..249
      NAME@229..233 "date"
      WHITESPACE@233..242 "         "
      EQ@242..243 "="
      WHITESPACE@243..244 " "
      LITERAL@244..248
        INTEGER@244..248 "2002"
      COMMA@248..249 ","
    WHITESPACE@249..254 "\n    "
    FIELD@254..272
      NAME@254..260 "volume"
      WHITESPACE@260..267 "       "
      EQ@267..268 "="
      WHITESPACE@268..269 " "
      LITERAL@269..271
        INTEGER@269..271 "97"
      COMMA@271..272 ","
    WHITESPACE@272..277 "\n    "
    FIELD@277..313
      NAME@277..282 "pages"
      WHITESPACE@282..290 "        "
      EQ@290..291 "="
      WHITESPACE@291..292 " "
      CURLY_GROUP@292..312
        L_CURLY@292..293 "{"
        WORD@293..301 "431--456"
        COMMA@301..302 ","
        WHITESPACE@302..303 " "
        WORD@303..311 "791--823"
        R_CURLY@311..312 "}"
      COMMA@312..313 ","
    WHITESPACE@313..318 "\n    "
    FIELD@318..342
      NAME@318..324 "langid"
      WHITESPACE@324..331 "       "
      EQ@331..332 "="
      WHITESPACE@332..333 " "
      CURLY_GROUP@333..341
        L_CURLY@333..334 "{"
        WORD@334..340 "german"
        R_CURLY@340..341 "}"
      COMMA@341..342 ","
    WHITESPACE@342..347 "\n    "
    FIELD@347..385
      NAME@347..357 "indextitle"
      WHITESPACE@357..360 "   "
      EQ@360..361 "="
      WHITESPACE@361..362 " "
      CURLY_GROUP@362..384
        L_CURLY@362..363 "{"
        WORD@363..365 "In"
        WHITESPACE@365..366 " "
        WORD@366..372 "Honore"
        WHITESPACE@372..373 " "
        WORD@373..383 "Salvatoris"
        R_CURLY@383..384 "}"
      COMMA@384..385 ","
    WHITESPACE@385..390 "\n    "
    FIELD@390..428
      NAME@390..400 "shorttitle"
      WHITESPACE@400..403 "   "
      EQ@403..404 "="
      WHITESPACE@404..405 " "
      CURLY_GROUP@405..427
        L_CURLY@405..406 "{"
        WORD@406..408 "In"
        WHITESPACE@408..409 " "
        WORD@409..415 "Honore"
        WHITESPACE@415..416 " "
        WORD@416..426 "Salvatoris"
        R_CURLY@426..427 "}"
      COMMA@427..428 ","
    WHITESPACE@428..433 "\n    "
    FIELD@433..614
      NAME@433..443 "annotation"
      WHITESPACE@443..446 "   "
      EQ@446..447 "="
      WHITESPACE@447..448 " "
      CURLY_GROUP@448..613
        L_CURLY@448..449 "{"
        WORD@449..450 "A"
        WHITESPACE@450..451 " "
        WORD@451..457 "German"
        WHITESPACE@457..458 " "
        WORD@458..465 "article"
        WHITESPACE@465..466 " "
        WORD@466..468 "in"
        WHITESPACE@468..469 " "
        WORD@469..470 "a"
        WHITESPACE@470..471 " "
        WORD@471..477 "French"
        WHITESPACE@477..478 " "
        WORD@478..486 "journal."
        WHITESPACE@486..487 " "
        WORD@487..492 "Apart"
        WHITESPACE@492..493 " "
        WORD@493..497 "from"
        WHITESPACE@497..498 " "
        WORD@498..502 "that"
        COMMA@502..503 ","
        WHITESPACE@503..504 " "
        WORD@504..505 "a"
        WHITESPACE@505..526 "\n                    "
        WORD@526..533 "typical"
        WHITESPACE@533..534 " "
        COMMAND@534..541
          COMMAND_NAME@534..541 "\\texttt"
        CURLY_GROUP@541..550
          L_CURLY@541..542 "{"
          WORD@542..549 "article"
          R_CURLY@549..550 "}"
        WHITESPACE@550..551 " "
        WORD@551..557 "entry."
        WHITESPACE@557..558 " "
        WORD@558..562 "Note"
        WHITESPACE@562..563 " "
        WORD@563..566 "the"
        WHITESPACE@566..567 " "
        COMMAND@567..574
          COMMAND_NAME@567..574 "\\texttt"
        CURLY_GROUP@574..586
          L_CURLY@574..575 "{"
          WORD@575..585 "indextitle"
          R_CURLY@585..586 "}"
        WHITESPACE@586..607 "\n                    "
        WORD@607..612 "field"
        R_CURLY@612..613 "}"
      COMMA@613..614 ","
    WHITESPACE@614..615 "\n"
    R_DELIM@615..616 "}"

